### ISSUE TITLE
Issue getting path from file located on SD card

### ISSUE THREAD

**Issue by j-roskopf** [iPaulPro/aFileChooser#54]
When using the aFileChooserExample application, I will select a file located on my SD card, and the resulting toast says "File Selected: null". Is there something I have to do to properly retrieve the path from the SD card?

**Comment by nicolabeghin** [comment]
I'm interested in this too. Works flawlessly but sometimes behaves like @j-roskopf wrote. I get 

> 11-05 15:27:47.820 I/MyApp(22768): onPause
> 11-05 15:27:47.860 I/Timeline(22768): Timeline: Activity_launch_request id:com.android.documentsui time:99395512
> 11-05 15:27:47.860 W/ResourceType(22768): CREATING STRING CACHE OF 44 bytes
> 11-05 15:27:50.250 W/MyApp(22768): content://com.android.externalstorage.documents/document/3037-6461%3Afile.txt

Maybe the 

> content://com.android.externalstorage.documents/document/3037-6461%3AChatStorage.sqlite
> is not treated in the right way?

I think the error lies in FileUtils.getPath:

```
public static String getPath(final Context context, final Uri uri) 
```

Any hint whatsoever?

thanks
nicola

**Comment by j-roskopf** [comment]
Hello. I believe I have found the error, in case you were still looking for it. In 4.4 and above, file access to the SD card is restricted, so I don't believe it to be a bug with the code

**Comment by ryanbehr** [comment]
Yes, this is correct. While they restrict, there are still ways to extract the file location from the URI provided by extracting everything after %3A and converting the rest of the URI to a real string (ex: %2F to forward slash). Then you can combine the path with the external SD card path:

```
Map<String, File> externalLocations = ExternalStorage.getAllStorageLocations();
File sdCard = externalLocations.get(ExternalStorage.SD_CARD);
File externalSdCard = externalLocations.get(ExternalStorage.EXTERNAL_SD_CARD);
File fullPath = new File(externalSdCard.getPath() + fixedPath);
```

So, if the path is null, try the alternate method to get the path.

**Comment by xandebianchi** [comment]
I did a modification in function getPath, FileUtils.java. It could solve the problem. I propose to solution and up it.

...
if ("primary".equalsIgnoreCase(type)) {
return Environment.getExternalStorageDirectory() + "/" + split[1];
} else {
String strPath;
       if (Environment.isExternalStorageRemovable()){
             strPath = System.getenv("EXTERNAL_STORAGE");
       } else{
             strPath = System.getenv("SECONDARY_STORAGE");
        if (strPath == null || strPath.length() == 0) {
             strPath = System.getenv("EXTERNAL_SDCARD_STORAGE");
         }
 }
...
